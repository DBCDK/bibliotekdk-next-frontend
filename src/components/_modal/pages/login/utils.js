import { LOGIN_MODE } from "@/components/_modal/pages/loanerform/LoanerForm";
import Translate from "@/components/base/translate";

/**
 * If we order a physical copy or a digital copy, we always show the loanerform
 * If the mode neither of the two modes above, we check if the library supports borrowerCheck
 * If not, we show the loginNotSupported modal, we can order stuff for these libraries, but we cannot login/validate via them
 * @param {string} mode
 * @returns {boolean}
 */
export const showLogin = (mode) => {
  return mode === LOGIN_MODE.ORDER_PHYSICAL && mode === LOGIN_MODE.SUBSCRIPTION;
};

export const isOrderPossible = ({ mode, branch }) => {
  if (!branch) return false;
  // QUICKFIX - .. to avoid api check ..  all public libraries have access to dda - no other
  // Færøerne skal ikke tjekkes
  if (mode === LOGIN_MODE.DDA) {
    return (
      branch?.agencyId?.startsWith("7") && branch.orderPolicy?.orderPossible
    );
  }
  return branch.orderPolicy?.orderPossible;
};

/**
 * Open login modal
 * @param {obj} modal
 * @param {string} title
 * @param {string} mode
 */
export function openLoginModal({
  modal,
  title = Translate({
    context: "header",
    label: "login",
  }),
  mode = LOGIN_MODE.PLAIN_LOGIN,
}) {
  console.log("utils", mode);
  modal.push("login", {
    title: title,
    mode: mode,
  });
}
