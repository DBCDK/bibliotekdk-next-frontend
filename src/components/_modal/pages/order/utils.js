import styles from "@/components/_modal/pages/order/Order.module.css";

/**
 *
 * @param {*} value
 * @param {*} valid
 * @param updateLoanerInfo
 * @param setMail
 */
export function onMailChange(value, valid, updateLoanerInfo, setMail) {
  valid.status &&
    updateLoanerInfo &&
    updateLoanerInfo({ userParameters: { userMail: value } });
  // update mail in state
  setMail({ value, valid });
}

/**
 *
 * @param pid
 * @param work
 * @param singleManifestation
 * @returns {*[]}
 */
export function getOrderPids(pid, work, singleManifestation) {
  // Work props
  const manifestations = work?.manifestations?.all;

  // Material by pid
  const material = manifestations?.filter(
    (manifestation) => manifestation.pid === pid
  );

  const materialsSameType = manifestations?.filter((manifestation) => {
    return (
      manifestation?.materialType === material?.materialType &&
      (!material?.accessTypes?.find(
        (accessType) => accessType?.display !== "fysisk"
      ) ||
        material?.access?.find((singleAccess) => singleAccess.loanIsPossible))
    );
  });

  let orderPids;
  if (singleManifestation) {
    orderPids = [pid];
  } else {
    orderPids = materialsSameType?.map((m) => m.pid);
  }
  return orderPids;
}

export function extractClassNameAndMessage(validated, failedSubmission) {
  // Get email messages (from validate object)
  const emailStatus = validated?.details?.hasMail?.status;
  const errorMessage = validated?.details?.hasMail?.message;

  // Check for email validation and email error messages
  const hasEmail = !!validated?.details?.hasMail?.status;

  // Email validation class
  const validClass =
    failedSubmission && !emailStatus ? styles.invalid : styles.valid;
  const customInvalidClass =
    failedSubmission && !emailStatus ? styles.invalidInput : "";

  // Set email input message if any
  const message = failedSubmission && errorMessage;

  const actionMessage =
    failedSubmission &&
    (validated?.details?.requireYear?.message ||
      (!hasEmail && validated?.details?.hasMail?.message));

  const invalidClass = actionMessage ? styles.invalid : "";

  return {
    emailStatus,
    errorMessage,
    hasEmail,
    validClass,
    customInvalidClass,
    message,
    actionMessage,
    invalidClass,
  };
}
