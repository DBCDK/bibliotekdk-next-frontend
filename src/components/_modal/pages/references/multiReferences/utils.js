/**
 * Splits the list materialsMissingEdition into two lists:
 * - newPeriodicaFiltered: list of periodica materials, for which we cannot create references
 * - listWithoutPeriodica: list of remaining materials, for which we can create references
 * @param {Array<Object>} materialsMissingEdition
 * @returns {Object}
 */
export const splitList = (materialsMissingEdition) => {
  const periodicaManifestations = [];
  const nonPeriodicaManifestations = materialsMissingEdition.map((item) => {
    const manifestations = item.manifestations;
    // If 1 option, select it
    if (manifestations.length === 1) {
      const singleManifestation = manifestations[0];
      if (singleManifestation.workTypes?.[0] === "PERIODICA") {
        //TODO What about journals??
        // Periodica -
        periodicaManifestations.push(singleManifestation);
        return {
          ...item,
          toFilter: true,
        };
      }
      return {
        ...item,
        chosenPid: singleManifestation.pid,
      };
    } else return item;
  });
  return {
    periodicaManifestations,
    nonPeriodicaManifestations,
  };
};
