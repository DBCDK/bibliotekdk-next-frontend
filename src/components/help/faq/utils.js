import orderBy from "lodash/orderBy";
import groupBy from "lodash/groupBy";
import get from "lodash/get";

import Translate from "@/components/base/translate";

/**
 * function to sort data for accordion structure
 *
 * @param {array} data
 *
 * @returns {array}
 */
export function sortData(data) {
  data = data.map((d) => {
    return { title: d.title, content: get(d, "body.value", "") };
  });

  // sort faq alfabetic order
  return orderBy(data, ["title"], ["asc"]);
}

/**
 * function to group data for accordion structure
 *
 * @param {array} data
 *
 * @returns {object}
 */
export function groupData(data) {
  const fallback = Translate({ context: "help", label: "faq-group-other" });

  const groups = groupBy(
    data,
    (e) => e?.fieldTags[0]?.entity?.entityLabel || fallback
  );

  return groups;
}

/**
 * function to group and sort data for accordion structure
 *
 * @param {array} data
 *
 * @returns {array}
 */
export function groupSortData(data) {
  const fallback = Translate({ context: "help", label: "faq-group-other" });

  const groups = groupBy(
    data,
    (e) => e?.fieldTags[0]?.entity?.entityLabel || fallback
  );

  const sortedGroups = {};
  Object.keys(groups).forEach((key) => {
    sortedGroups[key] = sortData(groups[key]);
  });

  return sortedGroups;
}
