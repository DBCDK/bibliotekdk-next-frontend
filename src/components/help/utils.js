/**
 * Parse helptexts by groups
 * @param helpTexts
 * @return {{}}
 *  eg. {SÃ¸gning:[{id:25, title:fisk}. {id:1,title:hest}]}
 */
export function helpTextParseMenu(helpTexts) {
  // sort helptexts by group
  const structuredHelpTexts = {};
  let element = {};
  let group;

  helpTexts.forEach((helptext, idx) => {
    element = setGroupElement(helptext);
    group = helptext.fieldHelpTextGroup;
    if (structuredHelpTexts[group]) {
      structuredHelpTexts[group].push(element);
    } else {
      structuredHelpTexts[group] = [element];
    }
  });
  return structuredHelpTexts;
}

/**
 * Defines an element in a help group
 * @param heltptext
 * @return {{id: (number|string|*), title}}
 */
function setGroupElement(heltptext) {
  return {
    id: heltptext.nid,
    title: heltptext.title,
  };
}
