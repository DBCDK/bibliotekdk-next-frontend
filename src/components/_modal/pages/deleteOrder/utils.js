import * as orderMutations from "@/lib/api/order.mutations";
import Translate from "@/components/base/translate";

export async function handleCancelOrder(orderId, agencyId, orderMutation) {
  try {
    return orderMutation.post(
      orderMutations.cancelOrder({
        orderId,
        agencyId,
      })
    );
  } catch (e) {
    console.log("could not delete orderq", e);

    return {
      error: `Could not delete order with orderId ${orderId} and agencyId ${agencyId} due to error ${e}`,
    };
  }
}

/**
 * Opens delete order modal
 * @param {*} param0.modal
 * @param {boolean} param0.mobile
 * @param {Date} param0.pickUpExpiryDate
 * @param {string} param0.id
 * @param {string} param0.agencyId
 * @param {string} param0.orderMutation
 * @returns success status, error message
 */
export function onClickDelete({
  modal,
  mobile,
  pickUpExpiryDate,
  id,
  agencyId,
  orderMutation,
  onCloseModal,
}) {
  modal.push("deleteOrder", {
    label: Translate({ context: "profile", label: "delete-order" }),
    mobile: mobile,
    isReadyToPickup: !!pickUpExpiryDate,
    orderId: id,
    agencyId: agencyId,
    orderMutation: orderMutation,
    onClose: onCloseModal,
  });
}
