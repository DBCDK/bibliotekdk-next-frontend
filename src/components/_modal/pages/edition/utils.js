import { getIsPeriodicaLike } from "@/lib/utils";

/**
 * Get a coverimage to use from given manifestations - from moreinfo OR default cover service
 * @param manifestations
 * @returns {{detail}|{detail: null}}
 */
export function editionCover(manifestations) {
  const manifestationWithCover = manifestations?.find(
    (manifestation) => manifestation?.cover.detail
  );
  return manifestationWithCover
    ? { detail: manifestationWithCover.cover.detail }
    : { detail: null };
}

export function inferAccessTypes(
  work,
  periodicaForm,
  initialPickupBranch,
  manifestationsBeforeCheck = null
) {
  const manifestations =
    manifestationsBeforeCheck || work?.manifestations?.all || [];

  const isArticle = !!work?.workTypes?.find(
    (workType) => workType.toLowerCase() === "article"
  );

  const isPeriodicaLike = getIsPeriodicaLike(
    work?.workTypes,
    work?.materialTypes
  );
  const isArticleRequest =
    !!periodicaForm?.titleOfComponent ||
    !!periodicaForm?.authorOfComponent ||
    !!periodicaForm?.pagination;

  const isDigitalCopy = !!manifestations?.find((m) => {
    return !!m?.access?.find((a) => a.issn);
  });

  const availableAsDigitalCopy =
    isDigitalCopy &&
    initialPickupBranch?.digitalCopyAccess &&
    (isPeriodicaLike ? isArticleRequest : true);

  const isPhysical = !!manifestations?.find((m) => {
    return m?.accessTypes?.find(
      (accessType) => accessType?.code === "PHYSICAL"
    );
  });
  const availableAsPhysicalCopy =
    isPhysical &&
    initialPickupBranch?.pickupAllowed &&
    initialPickupBranch?.orderPolicy?.orderPossible;

  const requireDigitalAccess = isDigitalCopy && !isPhysical;

  return {
    isArticle,
    isPeriodicaLike,
    isArticleRequest,
    isDigitalCopy,
    availableAsDigitalCopy,
    isPhysical,
    availableAsPhysicalCopy,
    requireDigitalAccess,
  };
}
