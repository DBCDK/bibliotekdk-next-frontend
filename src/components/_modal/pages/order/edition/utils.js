import { getIsPeriodicaLike } from "@/lib/utils";

export function inferCheckers(
  work,
  context,
  manifestations,
  initialPickupBranch
) {
  const isArticle = work?.workTypes?.find(
    (workType) => workType.toLowerCase() === "article"
  );
  const isPeriodicaLike = getIsPeriodicaLike(
    work?.workTypes,
    work?.materialTypes
  );
  const isArticleRequest =
    !!context?.periodicaForm?.titleOfComponent ||
    !!context?.periodicaForm?.authorOfComponent ||
    !!context?.periodicaForm?.pagination;
  const isDigitalCopy = !!manifestations?.find((m) =>
    m?.access?.find((entry) => entry.issn)
  );
  const availableAsDigitalCopy =
    initialPickupBranch?.pickupBranch?.digitalCopyAccess &&
    isDigitalCopy &&
    (isPeriodicaLike ? isArticleRequest : true);

  return {
    isArticle,
    isPeriodicaLike,
    isArticleRequest,
    isDigitalCopy,
    availableAsDigitalCopy,
  };
}
