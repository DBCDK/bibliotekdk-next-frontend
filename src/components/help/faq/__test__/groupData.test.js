import { groupData } from "../utils";
import data from "./mock/group.mock";

test("Group data", () => {
  const actual = groupData(data.faq.entities);
  expect(actual).toMatchSnapshot();
});
