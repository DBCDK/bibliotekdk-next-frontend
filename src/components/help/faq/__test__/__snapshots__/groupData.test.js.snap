// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group data 1`] = `
Object {
  "Andet": Array [
    Object {
      "body": Object {
        "value": "<p>This faq goes last in the group ...</p>",
      },
      "fieldTags": Array [],
      "langcode": Object {
        "value": "da",
      },
      "nid": 32,
      "promote": false,
      "title": "def",
    },
    Object {
      "body": Object {
        "value": "<p>This faq goes first in the group ...</p>",
      },
      "fieldTags": Array [],
      "langcode": Object {
        "value": "da",
      },
      "nid": 32,
      "promote": false,
      "title": "abc",
    },
  ],
  "Bestillinger": Array [
    Object {
      "body": Object {
        "value": "<p>This faq goes first in the group ...</p>",
      },
      "fieldTags": Array [
        Object {
          "entity": Object {
            "entityLabel": "Bestillinger",
          },
        },
      ],
      "langcode": Object {
        "value": "en",
      },
      "nid": 30,
      "promote": false,
      "title": "abc",
    },
  ],
  "Søgninger": Array [
    Object {
      "body": Object {
        "value": "<p>This faq goes last in the group ...</p>",
      },
      "fieldTags": Array [
        Object {
          "entity": Object {
            "entityLabel": "Søgninger",
          },
        },
      ],
      "langcode": Object {
        "value": "en",
      },
      "nid": 31,
      "promote": true,
      "title": "ghi",
    },
    Object {
      "body": Object {
        "value": "<p>This faq goes second in the group ...</p>",
      },
      "fieldTags": Array [
        Object {
          "entity": Object {
            "entityLabel": "Søgninger",
          },
        },
      ],
      "langcode": Object {
        "value": "en",
      },
      "nid": 31,
      "promote": true,
      "title": "def",
    },
    Object {
      "body": Object {
        "value": "<p>This faq goes first in the group...</p>",
      },
      "fieldTags": Array [
        Object {
          "entity": Object {
            "entityLabel": "Søgninger",
          },
        },
      ],
      "langcode": Object {
        "value": "en",
      },
      "nid": 31,
      "promote": true,
      "title": "abc",
    },
  ],
}
`;
